File: project-euler/square_digit_chain.py
from numba import njit
from tqdm import tqdm


@njit
def sum_square_digits(n):
  sum_squares = 0
  while n != 0:
    sum_squares += (n % 10) ** 2
    n //= 10
  return sum_squares


@njit
def digit_chain_recursive(n):
  if n == 0 or n == 1 or n == 89:
    return n
  else:
    return digit_chain_recursive(sum_square_digits(n))


@njit
def digit_chain_iterative(n):
  while True:
    if n == 0 or n == 1 or n == 89:
      return n
    else:
      n = sum_square_digits(n)


def digit_chain_iterative_with_dict(n):
  visited = dict({})
  while True:
    if n == 0 or n == 1 or n == 89:
      return n
    else:
      if n in visited:
        n = visited[n]
      else:
        visited[n] = sum_square_digits(n)


ends_with_89 = 0
ends_with_1 = 0
for i in tqdm(range(10**7)):
  if digit_chain_iterative(i + 1) == 1:
    ends_with_1 += 1
  else:
    ends_with_89 += 1

print(f"{ends_with_1}, {ends_with_89}")

File: project-euler/square_digit_chain.c
#include <math.h>
#include <stdio.h>

int sum_square_digits(int n) {
  int sum_squares = 0;
  while (n != 0) {
    sum_squares += pow((n % 10), 2);
    n /= 10;
  }
  return sum_squares;
}

int digit_chain_recursive(int n) {
  if (n == 0 || n == 1 || n == 89) {
    return n;
  } else {
    return digit_chain_recursive(sum_square_digits(n));
  }
}

int main() {
  int ends_with_1 = 0;
  int ends_with_89 = 0;
  for (int i = 0; i < 10e7; i++) {
    if (i % 1000000 == 0) {
      printf("%.2f %% done \n", (i / 10e7) * 100);
    }
    int ans = digit_chain_recursive(i);
    if (ans == 1) {
      ends_with_1 += 1;
    } else if (ans == 89) {
      ends_with_89 += 1;
    }
  }
  printf("%d -  %d", ends_with_1, ends_with_89);
}
